doc/developer/design/20210413_source_sink_resource_sharing.md

```diff
@@ -0,0 +1,269 @@
+
+  * Multiple instances of a Kafka source would share the underlying Kafka
+    consumer.
+  * Multiple instances of a file source would share the underlying file handler
+    and filesystem event listener.
+  * Multiple instances of a `TAIL` sink would share the underlying consolidation
+    operator.
+
+The proposal here is to teach Materialize to perform this resource sharing
+automatically and dynamically as the catalog evolves.
+
+## Goals
+
+There are three goals here:
+
+  * To raise awareness of the general problem of steady-state inefficiency.
+  * To build consensus that dynamic resource sharing is the general solution to
+    the problem.
+  * To commit to using dynamic resource sharing to permit unmaterialized
+    PostgreSQL sources.
+
+The last goal is in fact the motivating goal. Taken in isolation, supporting
+unmaterialized PostgreSQL sources does not clearly justify the complexities of
+dynamic resource sharing. But taken altogether, I think it becomes clear that
+there is a general problem here that warrants a general solution.
+
+## Non-goals
+
+The design here does not even begin to address the implementation of dynamic
+resource sharing. Suffice it to say that the implementations will be nontrivial.
+We'll need separate implementation design docs for each source and sink type.
+
+## Description
+
+### Background
+
+Consider the following DDL statements in Materialize today:
+
+```sql
+CREATE SOURCE kafka_src FROM KAFKA BROKER '...' TOPIC 'top' FORMAT AVRO ... ;
+CREATE MATERIALIZED VIEW view1 AS SELECT col1 FROM kafka_src;
+CREATE MATERIALIZED VIEW view2 AS SELECT col2 FROM kafka_src;
+```
+
+`view1` and `view2` will *each* instantiate `kafka_src`, resulting in
+two separate Kafka consumers that read from the `top` topic:
+
+```
+       kafka_src/1                    kafka_src/2
+  +---------------------+        +---------------------+
+  | librdkafka consumer |        | librdkafka consumer |
+  +---------------------+        +---------------------+
+  |    Avro decoding    |        |    Avro decoding    |
+  +---------------------+        +---------------------+
+           |                               |
+         view1                           view2
+```
+
+In the steady state, this is quite wasteful. We're streaming the records from
+`top` over the network twice and decoding the Avro in each record twice. We'd
+prefer to run just one Kafka consumer that feeds both views:
+
+```
+               +---------------------+
+               | librdkafka consumer |
+               +---------------------+
+               |    Avro decoding    |
+               +---------------------+
+             /                        \
+            /                           \
+      kafka_src/1                     kafka_src/2
+          |                                |
+        view1                            view2
+```
+
+### Today's workaround
+
+If you know the complete set of columns you care about ahead of time, you can
```

Comment:
I think there is a more fundamental reason than just ease-of-use for why this workaround is unsatisfying: it only works when the cardinalities of all the downstream views are the same.

For example, consider the following DDL fragment:

```
CREATE SOURCE src FROM KAFKA BROKER '...' TOPIC 'top' FORMAT AVRO;
CREATE MATERIALIZED VIEW v1 AS SELECT sum(col1) FROM src;
CREATE MATERIALIZED VIEW v2 AS SELECT col2 FROM src;
```

I don't believe there is any intermediate view you could materialize here that would avoid materializing more than you want (e.g., all the `col1`s).